src/graph/optimizer/rule/PushLimitDownIndexScanRule.cpp

```diff
@@ -0,0 +1,85 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "graph/optimizer/rule/PushLimitDownIndexScanRule.h"
+
+#include "common/expression/BinaryExpression.h"
+#include "common/expression/ConstantExpression.h"
+#include "common/expression/Expression.h"
+#include "common/expression/FunctionCallExpression.h"
+#include "common/expression/LogicalExpression.h"
+#include "common/expression/UnaryExpression.h"
+#include "graph/optimizer/OptContext.h"
+#include "graph/optimizer/OptGroup.h"
+#include "graph/planner/plan/PlanNode.h"
+#include "graph/planner/plan/Query.h"
+#include "graph/visitor/ExtractFilterExprVisitor.h"
+
+using nebula::graph::IndexScan;
+using nebula::graph::Limit;
+using nebula::graph::PlanNode;
+using nebula::graph::Project;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> PushLimitDownIndexScanRule::kInstance =
+    std::unique_ptr<PushLimitDownIndexScanRule>(new PushLimitDownIndexScanRule());
+
+PushLimitDownIndexScanRule::PushLimitDownIndexScanRule() { RuleSet::QueryRules().addRule(this); }
+
+const Pattern &PushLimitDownIndexScanRule::pattern() const {
+  static Pattern pattern = Pattern::create(
+      graph::PlanNode::Kind::kLimit,
+      {Pattern::create(graph::PlanNode::Kind::kProject,
+                       {Pattern::create(graph::PlanNode::Kind::kIndexScan, {}, true)})});
+  return pattern;
+}
+
+StatusOr<OptRule::TransformResult> PushLimitDownIndexScanRule::transform(
+    OptContext *octx, const MatchedResult &matched) const {
+  auto limitGroupNode = matched.node;
+  auto projGroupNode = matched.dependencies.front().node;
+  auto indexScanGroupNode = matched.dependencies.front().dependencies.front().node;
+
+  const auto limit = limitGroupNode->node()->asNode<const Limit>();
+  const auto proj = projGroupNode->node()->asNode<const Project>();
+  const auto indexScan = indexScanGroupNode->node()->asNode<const IndexScan>();
+
+  int64_t limitRows = limit->offset() + limit->count();
```

Comment:
Could you give me a comment? What does "offset" mean? 
for example `limit 5 offset 3` , Starting at line 3, continue to get 5 rows. the total limit is 5.

Reply:
I think `offset` is not easy to handle in IndexScan

Reply:
The sentence `limit 3, 5` means get 5 elements from the third element.

Reply:
@bright-starry-sky do we use `offset` to paging? Cursor will be a better choice.

Reply:
> @bright-starry-sky do we use `offset` to paging? Cursor will be a better choice.

No, we did not use `offset`, the current approach is similar to a cursor.

Reply:
> @bright-starry-sky do we use `offset` to paging? Cursor will be a better choice.

please ignore the concept of `offset`. It has nothing to do with index scanning.